lib/SPIRV/libSPIRV/SPIRVValue.h

```diff
@@ -146,69 +146,66 @@ template <spv::Op OC> class SPIRVConstantBase : public SPIRVValue {
   SPIRVConstantBase(SPIRVModule *M, SPIRVType *TheType, SPIRVId TheId,
                     uint64_t TheValue)
       : SPIRVValue(M, 0, OC, TheType, TheId) {
-    Union.UInt64Val = TheValue;
-    recalculateWordCount();
-    validate();
+    setWords(&TheValue);
   }
   // Incomplete constructor for AP integer constant
   SPIRVConstantBase(SPIRVModule *M, SPIRVType *TheType, SPIRVId TheId,
-                    llvm::APInt &TheValue);
+                    const llvm::APInt &TheValue);
   // Complete constructor for float constant
   SPIRVConstantBase(SPIRVModule *M, SPIRVType *TheType, SPIRVId TheId,
                     float TheValue)
       : SPIRVValue(M, 0, OC, TheType, TheId) {
-    Union.FloatVal = TheValue;
-    recalculateWordCount();
-    validate();
+    setWords(reinterpret_cast<uint64_t *>(&TheValue));
   }
   // Complete constructor for double constant
   SPIRVConstantBase(SPIRVModule *M, SPIRVType *TheType, SPIRVId TheId,
                     double TheValue)
       : SPIRVValue(M, 0, OC, TheType, TheId) {
-    Union.DoubleVal = TheValue;
-    recalculateWordCount();
-    validate();
+    setWords(reinterpret_cast<uint64_t *>(&TheValue));
   }
   // Incomplete constructor
   SPIRVConstantBase() : SPIRVValue(OC), NumWords(0) {}
-  uint64_t getZExtIntValue() const { return Union.UInt64Val; }
-  float getFloatValue() const { return Union.FloatVal; }
-  double getDoubleValue() const { return Union.DoubleVal; }
+  uint64_t getZExtIntValue() const { return getValue(); }
+  float getFloatValue() const { return static_cast<float>(getValue()); }
```

Comment:
Not sure if this is correct: `static_cast` will perform conversion operation ([cppreference](https://en.cppreference.com/w/cpp/language/static_cast) - see `1)`, [godbolt](https://godbolt.org/z/s4sn9b)), while you are interested in `bit_cast` (C++20) here.

Other possible options:
```c++
  return reinterpret_cast<float&>(getValue());
  // 6) from https://en.cppreference.com/w/cpp/language/reinterpret_cast
  // breaks type aliasing rules
```

```c++
  float res;
  std::memcpy(res, Words.data(), sizeof(res));
  return res;
  // this is a valid (non-exploiting a UB) option, which is expected to be optimized by compilers down to bitcast without actual memory copy
```

```c++
  union temp {uint64_t ivalue; float fvalue; };
  temp value;
  value.ivalue = getValue();
  return value.fvalue; // this is a UB, even though it works everywhere as expected
  // From https://en.cppreference.com/w/cpp/language/union
  // It's undefined behavior to read from the member of the union that wasn't most recently written.
  // Many compilers implement, as a non-standard language extension, the ability to read inactive members of a union.
```

Reply:
Thank you very much for the tips. I thought about another solution, but static_cast unexpectedly worked even for cast from unit64 to float.